### PR TITLE
API 4537 composite token list or search

### PR DESCRIPTION
**Pull request by that-guy-is-it**

### CODE REVIEW

File: patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java [hunk 1]
```diff
@@ -35,6 +35,11 @@ private static Ids ids() {
                 .system("https://veteran.apps.va.gov/appointments/v1")
                 .code("202008211400983000082100000000000000")
                 .build())
+        .queestionnaireResponseSecondaryMetaTag(
+            Ids.MetaTag.builder()
```
Comment (jblefkove-va) [inline patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java:39]: Seems like `questionnaireResponseMetaTag` and this should be combined into one class, similar to `UsageContextTypeValue` above

File: patient-generated-data-tests/src/test/java/gov/va/api/health/patientgenerateddata/tests/QuestionnaireResponseIT.java [hunk 2]
```diff
@@ -155,6 +155,47 @@ void search_tag_codeWithNoSystem() {
...
+    String secondaryTagCode =
+        systemDefinition().ids().queestionnaireResponseSecondaryMetaTag().code();
+    String secondaryTagSystem =
+        systemDefinition().ids().queestionnaireResponseSecondaryMetaTag().system();
+    String query =
+        String.format(
+            "?_tag=%s,%s", tagSystem + "|" + tagCode, secondaryTagSystem + "|" + secondaryTagCode);
+    var response = doGet("application/json", "QuestionnaireResponse" + query, 200);
+    QuestionnaireResponse.Bundle bundle = response.expectValid(QuestionnaireResponse.Bundle.class);
+    assertThat(bundle.entry()).hasSizeGreaterThan(0);
```
Comment (jblefkove-va) [inline patient-generated-data-tests/src/test/java/gov/va/api/health/patientgenerateddata/tests/QuestionnaireResponseIT.java:196]: Is it expecting 2?

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java [hunk 3]
```diff
@@ -28,6 +33,34 @@ static String addTerminators(String str) {
...
 
+  /**
+   * Takes a set of values to read from the database, and returns any values within the database
+   * that contain a given part of the set.
+   */
+  static <E> Specification<E> selectLikeInList(String fieldName, Set<String> values) {
+    if (values == null || values.isEmpty()) {
+      return null;
+    }
+    ArrayList<String> list = new ArrayList<>(values);
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:44]: This list doesn't seem necessary?

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java [hunk 4]
```diff
@@ -28,6 +33,34 @@ static String addTerminators(String str) {
     return "|" + str + "|";
   }
 
+  /**
+   * Takes a set of values to read from the database, and returns any values within the database
+   * that contain a given part of the set.
+   */
+  static <E> Specification<E> selectLikeInList(String fieldName, Set<String> values) {
+    if (values == null || values.isEmpty()) {
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:41]: You can use `CollectionUtils.isEmpty()`

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java [hunk 5]
```diff
@@ -28,6 +33,34 @@ static String addTerminators(String str) {
...
+    }
+    return (root, criteriaQuery, criteriaBuilder) -> {
+      List<Predicate> predicates =
+          list.stream()
+              .map(
+                  val ->
+                      criteriaBuilder.like(
+                          criteriaBuilder.lower(root.get(fieldName)),
+                          "%" + addTerminators(val).toLowerCase(Locale.ENGLISH) + "%"))
+              .collect(toList());
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:59]: Can you convert the stream directly to an array?

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java [hunk 6]
```diff
@@ -28,6 +33,34 @@ static String addTerminators(String str) {
...
+  static <E> Specification<E> selectLikeInList(String fieldName, Set<String> values) {
+    if (values == null || values.isEmpty()) {
+      return null;
+    }
+    ArrayList<String> list = new ArrayList<>(values);
+    if (list.size() == 1) {
+      return (root, criteriaQuery, criteriaBuilder) ->
+          criteriaBuilder.like(
+              criteriaBuilder.lower(root.get(fieldName)),
+              "%" + addTerminators(list.get(0)).toLowerCase(Locale.ENGLISH) + "%");
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:49]: `Iterables.getOnlyElement()`

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java [hunk 7]
```diff
@@ -84,14 +117,11 @@ public static String useContextValueJoin(Questionnaire questionnaire) {
...
-    if (isBlank(value)) {
-      return null;
-    }
-    return (Specification<EntityT>)
-        (root, criteriaQuery, criteriaBuilder) ->
-            criteriaBuilder.like(
-                criteriaBuilder.lower(root.get(fieldName)),
-                "%" + addTerminators(value).toLowerCase(Locale.ENGLISH) + "%");
+    var values =
+        Stream.of(request.getParameter(parameterName()).split("\\s*,\\s*"))
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:121]: Hmm, use `Splitter.on(",").trimResults()` so that the leading/trailing whitespace is also stripped

File: patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java (hunk 1 above)
Comment (that-guy-is-it) [reply patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java:39]: Hmm but if the idea is to have two full system and code pairs, wouldn't putting it in one class be counter productive? Or are you suggesting `questionnaireResponseMetaTag` contain a list of `MetaTag`?

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingTest.java [hunk 8]
```diff
@@ -28,6 +31,19 @@ private static Questionnaire _questionnaire(
         .build();
   }
 
+  @Test
+  void selectLikeInList() {
```
Comment (jblefkove-va) [inline patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingTest.java:35]: this should no longer be necessary

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java (hunk 7 above)
Comment (that-guy-is-it) [reply patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:121]: `Splitter.on(",").trimResults()` or `Splitter.on("\\s*,\\s*).trimResults()`?

File: patient-generated-data-tests/src/test/java/gov/va/api/health/patientgenerateddata/tests/QuestionnaireResponseIT.java [hunk 9]
```diff
@@ -155,6 +155,47 @@ void search_tag_codeWithNoSystem() {
...
+    bundle = response.expectValid(QuestionnaireResponse.Bundle.class);
+    assertThat(bundle.entry()).isEmpty();
+    query = "?_tag=|,";
+    response = doGet("application/json", "QuestionnaireResponse" + query, 200);
+    bundle = response.expectValid(QuestionnaireResponse.Bundle.class);
+    assertThat(bundle.entry()).isEmpty();
+  }
+
+  @Test
+  void search_tag_multipleSystemsAndCodes() {
```
Comment (jblefkove-va) [inline patient-generated-data-tests/src/test/java/gov/va/api/health/patientgenerateddata/tests/QuestionnaireResponseIT.java:182]: there should also be an IT for composite-mapping csv

File: patient-generated-data-tests/src/test/java/gov/va/api/health/patientgenerateddata/tests/QuestionnaireResponseIT.java [hunk 10]
```diff
@@ -155,6 +155,47 @@ void search_tag_codeWithNoSystem() {
     assertThat(bundle.entry()).isEmpty();
   }
 
+  @Test
+  void search_tag_empty() {
+    assumeEnvironmentIn(
+        Environment.LOCAL, Environment.QA, Environment.STAGING, Environment.STAGING_LAB);
+    // , Environment.LAB):
+    String query = "?_tag=,";
```
Comment (jblefkove-va) [inline patient-generated-data-tests/src/test/java/gov/va/api/health/patientgenerateddata/tests/QuestionnaireResponseIT.java:163]: Now that we have unit tests that can execute a whole specification, can these cases move down into unit tests instead?

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    request.setProtocol(u.getScheme());
+    request.setServerPort(u.getPort());
+    u.getQueryParams()
+        .entrySet()
+        .forEach(e -> request.addParameter(e.getKey(), e.getValue().toArray(new String[0])));
+    return request;
+  }
+
+  @Test
+  void doIt() {
```
Comment (jblefkove-va) [inline patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java:49]: better name?

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    request.setProtocol(u.getScheme());
+    request.setServerPort(u.getPort());
+    u.getQueryParams()
+        .entrySet()
+        .forEach(e -> request.addParameter(e.getKey(), e.getValue().toArray(new String[0])));
+    return request;
+  }
+
+  @Test
+  void doIt() {
```
Comment (apeirats-va) [inline patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java:57]: rename

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java (hunk 7 above)
Comment (jblefkove-va) [reply patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:121]: the former

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java (hunk 11 above)
Comment (that-guy-is-it) [reply patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java:49]: How about `findInRepository()`?

File: patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java (hunk 1 above)
Comment (jblefkove-va) [reply patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java:39]: Putting them in a list makes it difficult to tell which tag is for the appointment vs the system identifier. So one class with two `MetaTag` fields

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java (hunk 11 above)
Comment (jblefkove-va) [reply patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java:49]: `specificationFor_systemAndCode`?

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    jdbc.execute("create table app.foo (id varchar, value varchar)");
+    String tagJoin =
+        CompositeMapping.useContextValueJoin(_questionnaire("fizz", "buzz", "something", "else"));
+    FooEntity entity = FooEntity.builder().id("x").value(tagJoin).build();
+    repository.save(entity);
+    CompositeMapping<FooEntity> mapping =
+        CompositeMapping.<FooEntity>builder().parameterName("param").fieldName("value").build();
+    Specification<FooEntity> spec =
+        mapping.specificationFor(_requestFromUri("http://fizz.com?param=buzz$something,else"));
```
Comment (apeirats-va) [inline patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java:66]: What's going on here? Isn't this an invalid search? Shouldn't the search be `buzz$something|else`?

Comment (jblefkove-va) [reply patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java:66]: This coincidentally works, but it should be `something|else`

Comment (that-guy-is-it) [reply patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java:66]: Good catch

Comment (apeirats-va) [reply patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java:66]: Should this search be rejected with a 400? The comma strictly separates KVPs, right? So each value must include a `$`? Or do we just search for _any_ key with a value of `else`?

File: patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java [hunk 14]
```diff
@@ -34,11 +34,19 @@ private static Ids ids() {
...
-                .code("202008211400983000082100000000000000")
+        .questionnaireResponseMetas(
+            Ids.Metas.builder()
+                .appointmentTag(
+                    Ids.MetaTag.builder()
+                        .system(
+                            "https://api.va.gov/services/fhir/v0/r4/NamingSystem/va-appointment-identifier")
+                        .code("202008211400983000082100000000000000")
+                        .build())
+                .systemTag(
```
Comment (jblefkove-va) [inline patient-generated-data-tests/src/main/java/gov/va/api/health/patientgenerateddata/tests/SystemDefinitions.java:45]: this name is a bit confusing, I am going to rename this to `applicationTag`

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java [hunk 15]
```diff
@@ -84,14 +82,11 @@ public static String useContextValueJoin(Questionnaire questionnaire) {
...
-    }
-    return (Specification<EntityT>)
-        (root, criteriaQuery, criteriaBuilder) ->
-            criteriaBuilder.like(
-                criteriaBuilder.lower(root.get(fieldName)),
-                "%" + addTerminators(value).toLowerCase(Locale.ENGLISH) + "%");
+    var values =
+        Splitter.on(",").trimResults().splitToList(request.getParameter(parameterName())).stream()
+            .filter(StringUtils::isNotBlank)
+            .filter(str -> !str.equals("|"))
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/CompositeMapping.java:88]: `CompositeMapping` should also protect against lone `$`

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/SelectionUtils.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+import com.google.common.collect.Iterables;
+import java.util.Locale;
+import java.util.Set;
+import javax.persistence.criteria.Predicate;
+import lombok.Builder;
+import lombok.Value;
+import org.springframework.data.jpa.domain.Specification;
+
+@Value
+@Builder
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/SelectionUtils.java:14]: Value and Builder are not applicable here since this is not a model class. Since this is a utility class with just static methods, it should be `@UtilityClass`

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/SelectionUtils.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+public class SelectionUtils {
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/SelectionUtils.java:15]: Hmm, `MappingUtils`?

File: patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/SelectionUtils.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+  static String addTerminators(String str) {
+    return "|" + str + "|";
+  }
+
+  /**
+   * Takes a set of values to read from the database, and returns any values within the database
+   * that contain a given part of the set.
+   */
+  static <E> Specification<E> selectLikeInList(String fieldName, Set<String> values) {
+    if (values == null || isEmpty(values)) {
```
Comment (jblefkove-va) [inline patient-generated-data/src/main/java/gov/va/api/health/patientgenerateddata/SelectionUtils.java:26]: `isEmpty()` also handles the null check

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    request.setProtocol(u.getScheme());
+    request.setServerPort(u.getPort());
+    u.getQueryParams()
+        .entrySet()
+        .forEach(e -> request.addParameter(e.getKey(), e.getValue().toArray(new String[0])));
+    return request;
+  }
+
+  @Test
+  void specificationFor_systemAndCode() {
```
Comment (jblefkove-va) [inline patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/CompositeMappingJpaTest.java:57]: similar to the tag jpa test, there should be a test for the edge cases, including lone dollar-sign

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+    request.setProtocol(u.getScheme());
+    request.setServerPort(u.getPort());
+    u.getQueryParams()
+        .entrySet()
+        .forEach(e -> request.addParameter(e.getKey(), e.getValue().toArray(new String[0])));
+    return request;
+  }
+
+  @Test
+  void search_tag_edgeCases() {
```
Comment (jblefkove-va) [inline patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingJpaTest.java:49]: `specificationFor` is still the method being tested here, so this should be `specificationFor_edgeCases` to match the other methods

File: patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingTest.java [hunk 21]
```diff
@@ -35,6 +48,21 @@ void metadataTagJoin_Code() {
...
+        TokenListMapping.metadataTagJoin(
+            _questionnaireResponseCsv("clinics", "123", "somethingElse", "456"));
+    assertThat(join).contains(addTerminators("clinics|123"));
+    assertThat(join).contains(addTerminators("somethingElse|456"));
+    assertThat(join).contains(addTerminators("clinics|"));
+    assertThat(join).contains(addTerminators("somethingElse|"));
+    assertThat(join).contains(addTerminators("123"));
+    assertThat(join).contains(addTerminators("456"));
+    assertThat(join).doesNotContain(addTerminators("|123"));
+    assertThat(join).doesNotContain(addTerminators("|123"));
```
Comment (jblefkove-va) [inline patient-generated-data/src/test/java/gov/va/api/health/patientgenerateddata/TokenListMappingTest.java:63]: Should be `|456` ?

**Review by jblefkove-va — APPROVED** [review]
